File: lecture09/exercise09-A/template.cpp
// Exercise 09-A: Traversal

#include <iostream>
#include <queue>
#include <stack>
#include <string>
#include <vector>

using namespace std;

// Node structure

template <typename T>
struct Node {
    T	  value;
    Node *left;
    Node *right;
};

// Pre-defined Trees

Node<char> *AlgorithmTree() {
    return new Node<char>{'A',
    	new Node<char>{'L',
	    new Node<char>{'O',
		new Node<char>{'H', nullptr, nullptr},
		new Node<char>{'M', nullptr, nullptr}
	    },
	    new Node<char>{'R', nullptr, nullptr}
	},
	new Node<char>{'G',
	    new Node<char>{'I', nullptr, nullptr},
	    new Node<char>{'T', nullptr, nullptr}
	}
    };
}

// Traversal: BFS (Iterative)

template <typename T>
void bfs(Node<T> *root) {
}

// Traversal: DFS (Iterative)

template <typename T>
void dfs(Node<T> *root) {
}

// Traversal: DFS (Recursive)

template <typename T>
void dfs_recursive(Node<T> *root) {
}

// Main execution

int main(int argc, char *argv[]) {
    auto tree = AlgorithmTree();

    cout << "BFS:"; bfs(tree); cout << endl;
    cout << "DFS:"; dfs(tree); cout << endl;
    cout << "DFS:"; dfs_recursive(tree); cout << endl;

    return EXIT_SUCCESS;
}

File: lecture09/exercise09-A/template.py
#!/usr/bin/env python3

# Exercise 09-A: Traversal

from dataclasses import dataclass
from collections import deque

# Node structure

@dataclass
class Node:
    value:  str
    left:   'Node' = None
    right:  'Node' = None

# Pre-defined Trees

AlgorithmTree = \
    Node('A',
        Node('L',
            Node('O',
                Node('H'),
                Node('M')
            ),
            Node('R'),
        ),
        Node('G',
            Node('I'),
            Node('T'),
        ),
    )

# Traversal: BFS (Iterative)

def bfs(root: Node) -> None:
    pass

# Traversal: DFS (Iterative)

def dfs(root: Node) -> None:
    pass

# Traversal: DFS (Recursive)

def dfs_recursive(root: Node) -> None:
    pass

# Main Execution

def main():
    print(f'BFS:', end=''); bfs(AlgorithmTree); print()
    print(f'DFS:', end=''); dfs(AlgorithmTree); print()
    print(f'DFS:', end=''); dfs_recursive(AlgorithmTree); print()

if __name__ == '__main__':
    main()
